src/backend/x11/buffer.rs

```diff
@@ -0,0 +1,138 @@
+    Protocol(X11Error),
+
+    #[error("The Dmabuf had too many planes")]
+    TooManyPlanes,
+
+    #[error("Duplicating the file descriptors for the dmabuf handles failed")]
+    DupFailed(String),
+}
+
+impl From<X11Error> for CreatePixmapError {
+    fn from(e: X11Error) -> Self {
+        CreatePixmapError::Protocol(e)
+    }
+}
+
+impl From<ReplyOrIdError> for CreatePixmapError {
+    fn from(e: ReplyOrIdError) -> Self {
+        X11Error::from(e).into()
+    }
+}
+
+impl From<ConnectionError> for CreatePixmapError {
+    fn from(e: ConnectionError) -> Self {
+        X11Error::from(e).into()
+    }
+}
+
+#[derive(Debug)]
+pub struct Pixmap {
+    connection: Arc<XCBConnection>,
+    inner: u32,
+    width: u16,
+    height: u16,
+}
+
+impl Pixmap {
+    #[allow(dead_code)]
+    /// Creates a pixmap from a Dmabuf.
+    pub fn from_dmabuf(
+        connection: Arc<XCBConnection>,
+        window: &Window,
+        dmabuf: &Dmabuf,
+    ) -> Result<Pixmap, CreatePixmapError> {
+        if dmabuf.num_planes() > 4 {
+            return Err(CreatePixmapError::TooManyPlanes);
+        }
+
+        let xid = connection.generate_id()?;
+        let mut strides = dmabuf.strides();
+        let mut offsets = dmabuf.offsets();
+        let mut fds = Vec::new();
+
+        for handle in dmabuf.handles() {
+            // XCB closes the file descriptor after sending, so duplicate the file descriptor.
+            let fd: RawFd = fcntl::fcntl(
+                handle,
+                fcntl::FcntlArg::F_DUPFD_CLOEXEC(0), // Why is this 0?
+            )
+            .map_err(|e| CreatePixmapError::DupFailed(e.to_string()))?;
+
+            fds.push(RawFdContainer::new(fd))
+        }
+
+        connection.dri3_pixmap_from_buffers(
+            xid,
+            window.id(),
+            dmabuf.width() as u16,
+            dmabuf.height() as u16,
+            strides.next().unwrap(),
+            offsets.next().unwrap(),
+            strides.next().unwrap(),
+            offsets.next().unwrap(),
+            strides.next().unwrap(),
+            offsets.next().unwrap(),
+            strides.next().unwrap(),
+            offsets.next().unwrap(),
+            window.depth(),
+            32, // TODO: Stop hardcoding this
+            dmabuf.format().modifier.into(),
+            fds,
+        )?;
+
+        Ok(Pixmap {
+            connection,
+            inner: xid,
+            width: dmabuf.width() as u16,
+            height: dmabuf.height() as u16,
+        })
+    }
+
+    #[allow(dead_code)]
+    pub fn present(&self, window: &Window) -> Result<(), X11Error> {
+        self.connection.copy_area(
+            self.inner,
+            window.id(),
+            window.gc(),
+            0,
+            0,
+            0,
+            0,
+            self.width,
+            self.height,
+        )?;
+
+        Ok(())
+    }
+}
+
+impl Drop for Pixmap {
+    fn drop(&mut self) {
+        let _ = self.connection.free_pixmap(self.inner);
```

Comment:
In case you want to add a TODO for that (and you think that it is worth it): Next version of x11rb will bring `x11rb::protocol::xproto::PixmapWrapper` that wraps a pixmap and calls `free_pixmap` in `Drop`. So, exactly what you are doing here. It even ignores any errors, just like this code. :-)

It's funny how often I saw code like this already since these wrappers were added. I guess they really are useful...